rasterio/aws.py

```diff
@@ -0,0 +1,43 @@
+import os
+
+from rasterio._drivers import ConfigEnv
+from rasterio.five import configparser
+
+
+class Session(ConfigEnv):
+
+    def __init__(self, aws_access_key_id=None, aws_secret_access_key=None,
+                 aws_session_token=None, region_name=None,
+                 config_dir='~/.aws', **options):
+        # Set GDAL config options for AWS. Precedence order:
+        # 1. Constructor args.
+        # 2. Well-known environment variables.
+        # 3. Config files in ~/.aws/
```

Comment:
This is related to https://trac.osgeo.org/gdal/ticket/6420.

When you're on the AWS infrastructure with S3 permissions granted to the machine via IAM roles, there is no config file or env vars holding these credentials. When you're on an EC2, the awscli seems to be able to perform [an extra step](http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html) to get that info from the instance metadata.

If we could emulate the awscli logic here, we could bypass that limitation in GDAL.